### PR TITLE
Add a new ANTLR lexer for Painless suggestions

### PR DESCRIPTION
**Pull request by jdconrad**
This splits off just the required ANTLR lexer from (https://github.com/elastic/elasticsearch/pull/70356) for Painless suggestions. This lexer is different from the standard Painless lexer in that is supports types within the lexer from a PainlessLookup. This is possible because a PainlessLookup is always required to generate appropriate suggestions. This also makes the required state machines coming up much simpler to implement.

### CODE REVIEW

File: modules/lang-painless/build.gradle
Comment (stu-elastic) [inline modules/lang-painless/build.gradle:246]: When I run this path I'm getting some changes to files and untracked files.

```
git status
On branch jack-autolexer
Your branch is up to date with 'jack/autolexer'.

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        deleted:    src/main/antlr/SuggestLexer.tokens
        modified:   src/main/java/org/elasticsearch/painless/antlr/SuggestLexer.java

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        src/main/java/org/elasticsearch/painless/antlr/SuggestLexer.tokens
```

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java:17]: customized for painless suggestions?

File: modules/lang-painless/src/main/antlr/SuggestLexer.g4
Comment (stu-elastic) [inline modules/lang-painless/src/main/antlr/SuggestLexer.g4:9]: This is substantily similar to `PainlessLexer.g4`, did you consider factoring out the commonalities and importing them in?

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java:19]: Is this true?

File: modules/lang-painless/build.gradle
Comment (jdconrad) [reply modules/lang-painless/build.gradle:246]: Discussed with @stu-elastic and the regenSuggestLexer command is meant to be a dependency of regenSuggest which does not leave behind these additional files.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java:46]: Is this so that we can keep suggesting if we fail in `DOTID` or `DOTINTEGER`?

**Review by stu-elastic — APPROVED** [review]
LGTM after responding to comments, thanks for breaking this up.

File: modules/lang-painless/src/main/antlr/SuggestLexer.g4
Comment (jdconrad) [reply modules/lang-painless/src/main/antlr/SuggestLexer.g4:9]: I did, but given this is iteration one and there will likely be more changes required, I think it would be beneficial to hold off on any kind of merging until we have a more finalized view.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java:46]: Yes! We drop back to default mode so we can continue lexing.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java:17]: No, cleaning up now.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java
Comment (jdconrad) [reply modules/lang-painless/src/main/java/org/elasticsearch/painless/antlr/EnhancedSuggestLexer.java:19]: No, cleaning up now.

**Comment by jdconrad** [comment]
@stu-elastic Thanks for the review. Addressed all the comments and will commit as soon as CI passes.

**Comment by jdconrad** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Pull request merged**
